src/SqlAst/AvgReturnTypeExtension.php

```diff
@@ -25,8 +29,38 @@ public function getReturnType(FunctionCall $expression, QueryScope $scope): ?Typ
         }
 
         $argType = $scope->getType($args[0]);
-        $numberType = $argType->toNumber();
 
-        return TypeCombinator::addNull($numberType);
+        if ($argType->isNull()->yes()) {
+            return $argType;
+        }
+        $argType = TypeCombinator::removeNull($argType);
+
+        if ($argType instanceof UnionType) {
+            $newTypes = [];
+            foreach ($argType->getTypes() as $type) {
+                $newTypes[] = $this->convertToAvgType($type);
+            }
+            $newType = TypeCombinator::union(...$newTypes);
+        } else {
+            $newType = $this->convertToAvgType($argType);
+        }
+
+        return TypeCombinator::addNull($newType);
+    }
+
+    private function convertToAvgType(Type $argType): Type
+    {
+        if ($argType->isInteger()->yes()) {
+            return new IntersectionType([
+                new StringType(),
+                new AccessoryNumericStringType(),
+            ]);
+        }
+
+        if ($argType->isString()->yes()) {
+            return $argType->toFloat();
+        }
```

Comment:
I think this impl deserves test-cases for a "non-empty-string", "non-falsey-string" and "numeric-string"

Reply:
Hmm, I don't know how to coerce a column/field type to be any of those types explicitly, so I'm not sure how I would get those types of inputs to `avg` in this style of test. I know I can get a constant string that is non-empty or numeric, such as `SELECT avg('foo')` or `SELECT avg('123.45')`. Is that what you're looking for? Otherwise, if you have an example, that would be really helpful. Thanks!

Reply:
using CONCAT it should be possible to construct these strings, see https://github.com/staabm/phpstan-dba/blob/6b6e1ca1213fc9bcb4440a946da6b1302297edc1/src/SqlAst/ConcatReturnTypeExtension.php#L65-L76

Reply:
Thanks for the suggestion, this worked great. And just to confirm the expected output types:
```php
$result = $pdo->query('SELECT avg(concat("0"))');
var_dump($result->fetch());
$result = $pdo->query('SELECT avg(concat("foo"))');
var_dump($result->fetch());
```

```
array (size=2)
  'avg(concat("0"))' => float 0
  0 => float 0
array (size=2)
  'avg(concat("foo"))' => float 0
  0 => float 0
```